README
Catmandu::BibTeX - modules for working with bibtex data

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Catmandu::BibTeX

Changes
Revision history for Catmandu-BibTeX

{{$NEXT}}

0.21  2021-06-08 11:16:20 CEST
  - use double brackets to preserve case

0.20  2021-03-02 10:19:26 CET
  - update tests to work with latest BibTeX::Parser (page range divider is now
    long dash)

0.19  2021-01-25 13:14:08 CET
  - fix version for LaTeX::ToUnicode

0.18  2020-03-04 13:39:24 CET
  - fix travis build

0.17  2020-03-04 13:28:49 CET
  - fix travis build

0.16  2020-03-04 13:21:27 CET
  - translate latex field values to unicode
  - better author/editor field handling
  - author and fields are always an array

0.15  2019-12-20 08:43:00 CET
  - add fix latex_decode

0.14  2019-07-04 09:34:49 CEST
  - fix typo

0.13  2019-07-03 15:45:24 CEST
  - fix typo (keyword --> keywords) in exporter

0.12  2019-06-25 16:14:43 CEST
  - use Test2::V0 for test
  - travis-ci update

0.11  2019-01-15 13:29:22 CET
  - provide VERSION for all *.pm files

0.10  2019-01-15 13:26:40 CET
  - make exporter idempotent

0.08  2017-01-17 08:01:25 CET
  - support importing authors/editors as if,
    i.e. single authors will be SCALARs and multiple authors
    will be ARRAYs

0.07  2015-08-26 10:29:12 CEST
  - remove latex_encode, leave utf-8 untouched

0.06  2014-09-30 09:36:49 CEST
  - well, see 0.05

0.05  2014-09-29 18:16:59 CEST
  - moved to milla
  - tests + badges

0.04  2014-03-04
  - date expansion
  - default citekey starts at 1
  - document default citekey
  - fix roundtripping
  - add project links to build script

0.03  2013-06-24
  - BibTeX importer

0.02  2013-06-12
  - initial release


lib/Catmandu/Exporter/BibTeX.pm
package Catmandu::Exporter::BibTeX;

our $VERSION = '0.20';

use namespace::clean;
use Catmandu::Sane;
use Clone qw(clone);
use Moo;

with 'Catmandu::Exporter';

my $TAGS = [
    qw(
        abstract
        address
        articleno
        author
        booktitle
        chapter
        day
        edition
        editor
        eprint
        howpublished
        institution
        isbn
        issn
        journal
        keywords
        language
        location
        month
        note
        number
        organization
        pages
        publisher
        school
        series
        title
        type
        url
        doi
        volume
        year
        )
];

my $JOIN = {author => ' and ', editor => ' and ', language => ',',
    keywords => ',',};

sub add {
    my ($self, $orig_data) = @_;

    my $data = clone($orig_data);

    my $fh = $self->fh;

    my $type    = $data->{type}     || $data->{_type} || 'misc';
    my $citekey = $data->{_citekey} || $data->{_id}   || $self->count + 1;

    for my $tag (keys %$JOIN) {
        my $val = $data->{$tag};
        if ($val && ref($val) eq 'ARRAY') {
            $data->{$tag} = join $JOIN->{$tag}, @$val;
        }
    }

    print $fh "\@$type\{$citekey,\n";

    for my $tag (@$TAGS) {
        if (my $val = $data->{$tag}) {
            printf $fh "  %-12s = {{%s}},\n", $tag, $val;
        }
    }

    print $fh "}\n\n";
}

=head1 NAME

Catmandu::Exporter::BibTeX - a BibTeX exporter

=head1 SYNOPSIS

    use Catmandu::Exporter::BibTeX;

    my $exporter = Catmandu::Exporter::BibTeX->new(fix => 'myfix.txt');

    $exporter->add_many($arrayref);
    $exporter->add_many($iterator);
    $exporter->add_many(sub { });

    $exporter->add($hashref);

    $exporter->add({
     type    => 'book',
     _citekey => '389-ajk0-1',
     title    => 'the Zen of {CSS} design',
     author   => ['Dave Shea','Molley E. Holzschlag'],
     isbn     => '0-321-30347-4'
    });

    printf "exported %d objects\n" , $exporter->count;

=head1 DESCRIPTION

The BibTeX L<Catmandu::Exporter> requires as input a Perl hash (or a fix)
containing BibTeX fields and values as a string or array reference.

=head1 SUPPORTED FIELDS

Two special fields can be set in the Perl hash:

=over

=item C<type> or C<_type>

to describe the document type (article, book, ...). Set to 'misc' by default.

=item C<_citekey> or C<_id>

to describt the citation key. The next counter value (starting from 1) is used
by default.

=back

The following BibTeX fields are supported. All other fields are ignored.

    abstract
    address
    author
    booktitle
    chapter
    day
    edition
    editor
    eprint
    howpublished
    institution
    isbn
    issn
    journal
    keywords
    language
    location
    month
    note
    number
    organization
    pages
    publisher
    school
    series
    title
    type
    url
    doi
    volume
    year

=head1 SEE ALSO

Use L<Catmandu::Fix::expand_date> to expand a date field with year, month, and day
into the corresponding BibTeX fields.

=cut

1;


lib/Catmandu/Fix/expand_date.pm
package Catmandu::Fix::expand_date;

use Catmandu::Sane;
use Moo;

our $dateRegexp = qr{
    ^([0-9]{4})
        ( [:-] ([0-9]{1,2})
            ( [:-] ([0-9]{1,2}) )?
        )?
}x;

sub fix {
    my ($self, $data) = @_;
    if ($data->{date} && $data->{date} =~ $dateRegexp) {
        $data->{year}  = $1;
        $data->{month} = 1*$3 if $3;
        $data->{day}   = 1*$5 if $5;
    }
    $data;
}

=head1 NAME

Catmandu::Fix::expand_date - expand a date field into year, month, and date

=head1 SYNOPSIS

    # date is "2001-09-11"
    expand_date();
    # year is "2001", month is "9", day is "11"

=head1 DESCRIPTION

The date field is expanded if it contains a year, optionally followed by
numeric month and day, each separated by C<-> or C<:>.

=head1 SEE ALSO

For more elaborated date parsing see L<Date::Parse>.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.012.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4229"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Catmandu modules for working with BibTeX data",
  "dist_author" => [
    "Nicolas Steenlant C<< <nicolas.steenlant at ugent.be> >>"
  ],
  "dist_name" => "Catmandu-BibTeX",
  "dist_version" => "0.19",
  "license" => "perl",
  "module_name" => "Catmandu::BibTeX",
  "recursive_test_files" => 1,
  "requires" => {
    "BibTeX::Parser" => "1.03",
    "Catmandu" => "1.2002",
    "Clone" => 0,
    "LaTeX::Decode" => "0.05",
    "Module::Build" => "0.4229",
    "Moo" => 0,
    "perl" => "v5.10.1"
  },
  "test_requires" => {
    "Test2::V0" => "0.000122"
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.4229",
  "Test2::V0" => "0.000122"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
